python/cudf/cudf/core/frame.py

```diff
@@ -1246,18 +1246,24 @@ def fillna(
                 for key, value in value.items()
             }
 
-        copy_data = self._data.copy(deep=True)
-
-        for name in copy_data.keys():
+        filled_data = {}
+        for col_name, col in self._data.items():
             should_fill = (
-                name in value
-                and not libcudf.scalar._is_null_host_scalar(value[name])
+                col_name in value
+                and col.can_fillna()
+                and not libcudf.scalar._is_null_host_scalar(value[col_name])
             ) or method is not None
             if should_fill:
-                copy_data[name] = copy_data[name].fillna(value[name], method)
-        result = self._from_data(copy_data, self._index)
+                filled_data[col_name] = col.fillna(value[col_name], method)
+            else:
+                filled_data[col_name] = col.copy(deep=True)
 
-        return self._mimic_inplace(result, inplace=inplace)
+        return self._mimic_inplace(
+            self._from_data(
+                data=ColumnAccessor(filled_data), index=self._index
```

Comment:
The conversion is performed for you in `_from_data`. You only need to do it explicitly if you have MultiIndex columns since that won't be inferred by the ColumnAccessor constructor.
```suggestion
                data=filled_data, index=self._index
```